src/channel.c

```diff
@@ -1889,11 +1889,17 @@ ssize_t _libssh2_channel_read(LIBSSH2_CHANNEL *channel, int stream_id,
            return code. */
         if(channel->remote.eof || channel->remote.close)
             return 0;
-        else if(rc != LIBSSH2_ERROR_EAGAIN)
-            return 0;
 
-        /* if the transport layer said EAGAIN then we say so as well */
-        return _libssh2_error(session, rc, "would block");
+        /* If the transport layer said EAGAIN then we say so as well.
+           If there wasn't any error, it means the transport layer got
+           new data but it went to some other place and so we return
+           EAGAIN too. */
+        if((rc == LIBSSH2_ERROR_EAGAIN) || (rc == LIBSSH2_ERROR_NONE))
```

Comment:
`_libssh2_transport_read` returns `LIBSSH2_ERROR_NONE` if `session->socket_state == LIBSSH2_SOCKET_DISCONNECTED`.  Is retrying with EAGAIN the correct response to this?  Or should `_libssh2_transport_read` be changed to return `LIBSSH2_ERROR_SOCKET_DISCONNECT` in that case?

I'm not sure i understand 'it means the transport layer got new data but it went to some other place'.  What other place? If I'm reading the code right, `rc == LIBSSH2_ERROR_NONE` can happen if the first byte of the payload is 0 (see fullpacket).  What does that signify?